File: app/src/main/java/de/sneakpeek/data/MovieInfo.java
package de.sneakpeek.data;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.squareup.moshi.Json;

import java.util.List;

public class MovieInfo implements Parcelable {

    public final static Parcelable.Creator<MovieInfo> CREATOR = new Creator<MovieInfo>() {

        public MovieInfo createFromParcel(Parcel in) {
            MovieInfo instance = new MovieInfo();
            instance.adult = ((boolean) in.readValue((boolean.class.getClassLoader())));
            instance.backdropPath = ((String) in.readValue((String.class.getClassLoader())));
            instance.belongsToCollection = ((Object) in.readValue((Object.class.getClassLoader())));
            instance.budget = ((int) in.readValue((int.class.getClassLoader())));
            in.readList(instance.genres, (de.sneakpeek.data.MovieInfoGenre.class.getClassLoader()));
            instance.homepage = ((String) in.readValue((String.class.getClassLoader())));
            instance.id = ((int) in.readValue((int.class.getClassLoader())));
            instance.imdbId = ((String) in.readValue((String.class.getClassLoader())));
            instance.originalLanguage = ((String) in.readValue((String.class.getClassLoader())));
            instance.originalTitle = ((String) in.readValue((String.class.getClassLoader())));
            instance.overview = ((String) in.readValue((String.class.getClassLoader())));
            instance.popularity = ((double) in.readValue((double.class.getClassLoader())));
            instance.posterPath = ((String) in.readValue((String.class.getClassLoader())));
            in.readList(instance.productionCompanies, (de.sneakpeek.data.MovieInfoProductionCompany.class.getClassLoader()));
            in.readList(instance.productionCountries, (de.sneakpeek.data.MovieInfoProductionCountry.class.getClassLoader()));
            instance.releaseDate = ((String) in.readValue((String.class.getClassLoader())));
            instance.revenue = ((int) in.readValue((int.class.getClassLoader())));
            instance.runtime = ((int) in.readValue((int.class.getClassLoader())));
            in.readList(instance.spokenLanguages, (de.sneakpeek.data.MovieInfoSpokenLanguage.class.getClassLoader()));
            instance.status = ((String) in.readValue((String.class.getClassLoader())));
            instance.tagline = ((String) in.readValue((String.class.getClassLoader())));
            instance.title = ((String) in.readValue((String.class.getClassLoader())));
            instance.video = ((boolean) in.readValue((boolean.class.getClassLoader())));
            instance.voteAverage = ((double) in.readValue((double.class.getClassLoader())));
            instance.voteCount = ((int) in.readValue((int.class.getClassLoader())));
            return instance;
        }

        public MovieInfo[] newArray(int size) {
            return (new MovieInfo[size]);
        }

    };
    @Json(name = "adult") public boolean adult;
    @Json(name = "backdrop_path") public String backdropPath;
    @Json(name = "belongs_to_collection") public Object belongsToCollection;
    @Json(name = "budget") public int budget;
    @Json(name = "genres") public List<MovieInfoGenre> genres = null;
    @Json(name = "homepage") public String homepage;
    @Json(name = "id") public int id;
    @Json(name = "imdb_id") public String imdbId;
    @Json(name = "original_language") public String originalLanguage;
    @Json(name = "original_title") public String originalTitle;
    @Json(name = "overview") public String overview;
    @Json(name = "popularity") public double popularity;
    @Json(name = "poster_path") public String posterPath;
    @Json(name = "production_companies") public List<MovieInfoProductionCompany> productionCompanies = null;
    @Json(name = "production_countries") public List<MovieInfoProductionCountry> productionCountries = null;
    @Json(name = "release_date") public String releaseDate;
    @Json(name = "revenue") public int revenue;
    @Json(name = "runtime") public int runtime;
    @Json(name = "spoken_languages") public List<MovieInfoSpokenLanguage> spokenLanguages = null;
    @Json(name = "status") public String status;
    @Json(name = "tagline") public String tagline;
    @Json(name = "title") public String title;
    @Json(name = "video") public boolean video;
    @Json(name = "vote_average") public double voteAverage;
    @Json(name = "vote_count") public int voteCount;
    @Json(name = "credits") public MovieInfoCredits credits;

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(adult);
        dest.writeValue(backdropPath);
        dest.writeValue(belongsToCollection);
        dest.writeValue(budget);
        dest.writeList(genres);
        dest.writeValue(homepage);
        dest.writeValue(id);
        dest.writeValue(imdbId);
        dest.writeValue(originalLanguage);
        dest.writeValue(originalTitle);
        dest.writeValue(overview);
        dest.writeValue(popularity);
        dest.writeValue(posterPath);
        dest.writeList(productionCompanies);
        dest.writeList(productionCountries);
        dest.writeValue(releaseDate);
        dest.writeValue(revenue);
        dest.writeValue(runtime);
        dest.writeList(spokenLanguages);
        dest.writeValue(status);
        dest.writeValue(tagline);
        dest.writeValue(title);
        dest.writeValue(video);
        dest.writeValue(voteAverage);
        dest.writeValue(voteCount);
    }

    public int describeContents() {
        return 0;
    }

    public
    @Nullable
    String getDirector() {
        if (credits != null && credits.crew != null) {
            for (MovieInfoCrew crew : credits.crew) {
                if (crew.job.equalsIgnoreCase("director")) {
                    return crew.name;
                }
            }
        }

        return null;
    }
}

File: app/src/main/java/de/sneakpeek/data/MovieInfo.kt
package de.sneakpeek.data

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

data class MovieInfo(@Expose @SerializedName("adult") var adult: Boolean = false,
                     @Expose @SerializedName("genres") var genres: List<MovieInfoGenre>?,
                     @Expose @SerializedName("id") var id: Int = 0,
                     @Expose @SerializedName("budget") var budget: Int = 0,
                     @Expose @SerializedName("imdb_id") var imdb_id: String?,
                     @Expose @SerializedName("original_language") var original_language: String?,
                     @Expose @SerializedName("original_title") var original_title: String?,
                     @Expose @SerializedName("overview") var overview: String?,
                     @Expose @SerializedName("popularity") var popularity: Double = 0.toDouble(),
                     @Expose @SerializedName("poster_path") var poster_path: String?,
                     @Expose @SerializedName("production_companies") var productionCompanies: List<MovieInfoProductionCompany>?,
                     @Expose @SerializedName("release_date") var release_date: String?,
                     @Expose @SerializedName("runtime") var runtime: Int = 0,
                     @Expose @SerializedName("tagline") var tagline: String?,
                     @Expose @SerializedName("title") var title: String?,
                     @Expose @SerializedName("vote_average") var vote_average: Double = 0.toDouble(),
                     @Expose @SerializedName("vote_count") var vote_count: Int = 0,
                     @Expose @SerializedName("credits") var credits: MovieInfoCredits?) : Parcelable {

    var director: String = ""
        get() {
            return credits?.crew
                    ?.firstOrNull { it.job.equals("director", ignoreCase = true) }
                    ?.name
                    ?: ""
        }

    var writer: String = ""
        get() {
            return credits?.crew
                    ?.firstOrNull { it.job.equals("writer", ignoreCase = true) }
                    ?.name
                    ?: ""
        }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<MovieInfo> = object : Parcelable.Creator<MovieInfo> {
            override fun createFromParcel(source: Parcel): MovieInfo = MovieInfo(source)
            override fun newArray(size: Int): Array<MovieInfo?> = arrayOfNulls(size)
        }
    }

    constructor(source: Parcel) : this(
            1 == source.readInt(),
            source.createTypedArrayList(MovieInfoGenre.CREATOR),
            source.readInt(),
            source.readInt(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readDouble(),
            source.readString(),
            source.createTypedArrayList(MovieInfoProductionCompany.CREATOR),
            source.readString(),
            source.readInt(),
            source.readString(),
            source.readString(),
            source.readDouble(),
            source.readInt(),
            source.readParcelable<MovieInfoCredits>(MovieInfoCredits::class.java.classLoader)
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt((if (adult) 1 else 0))
        dest.writeTypedList(genres)
        dest.writeInt(id)
        dest.writeInt(budget)
        dest.writeString(imdb_id)
        dest.writeString(original_language)
        dest.writeString(original_title)
        dest.writeString(overview)
        dest.writeDouble(popularity)
        dest.writeString(poster_path)
        dest.writeTypedList(productionCompanies)
        dest.writeString(release_date)
        dest.writeInt(runtime)
        dest.writeString(tagline)
        dest.writeString(title)
        dest.writeDouble(vote_average)
        dest.writeInt(vote_count)
        dest.writeParcelable(credits, 0)
    }
}